src/test/sql/sql_to_ast_translator_test.cpp

```diff
@@ -473,4 +463,68 @@ TEST_F(SQLToASTTranslatorTest, Update) {
   EXPECT_EQ(*expressions[1]->alias(), "b");
 }
 
+TEST_F(SQLToASTTranslatorTest, MixedAggregateAndGroupBySelectList) {
+  /**
+   * Test:
+   *    - Select list can contain both GroupBy and Aggregate Columns in any order
+   *    - The order of GroupBy Columns in the SelectList can differ from the order in the GroupByList
+   */
+
+  const auto query = R"(SELECT
+                          table_b.a, SUM(table_c.a), table_a.b, table_b.b
+                        FROM
+                          table_a, table_b, table_c
+                        GROUP BY
+                          table_a.b, table_a.a, table_b.a, table_b.b)";
+
+  /**
+   * Expect this AST:
+   *
+   * [Projection] table_a.a, SUM(table_c.a), table_a.b, table_a.b
+   *  |_[Aggregate] SUM(table_c.a) GROUP BY [table_a.b, table_a.a, table_b.a, table_b.b]
+   *     |_[Cross Join]
+   *        |_[Cross Join]
+   *         |  |_[StoredTable] Name: 'table_b'
+   *         |  |_[StoredTable] Name: 'table_c'
+   *         |_[StoredTable] Name: 'table_a'
+   */
+
+  auto result = compile_query(query);
+
+  /**
+   * Assert the Projection
+   */
+  ASSERT_EQ(result->type(), ASTNodeType::Projection);
+  const auto projection_node = std::dynamic_pointer_cast<ProjectionNode>(result);
+  ASSERT_EQ(projection_node->column_expressions().size(), 4u);
+  EXPECT_COLUMN_EXPRESSION(projection_node->column_expressions()[0], ColumnID{2});
```

Comment:
Why do we introduce a load of new macros? This
* is longer than `EXPECT_EQ(projection_node->column_expressions()[0], ColumnID{2});`,
* does not print the actual and expected values (well, we could change that in `EXPECT_COLUMN_EXPRESSION`)
* does exactly the same, adding another layer of indirection that the user has to follow,
* adds another test macro that the user has to remember,
* and adds ambiguity to places where the `EXPECT_EQ` version is used.

Reply:
I introduce these to make tests clearly state the semantics they are testing (EXPECT_COLUMN_EXPRESSION), because that's really something our tests are lacking (mine are no exception). If I read `EXPECT_EQ(projection_node->column_expressions()[0], ColumnID{2});` I have no idea what's actually being tested here.

As to what they print on failure - you're right that's important but worth discussing in another issue? Prints when tables to not match are a catastrophe atm as well...

Reply:
I am afraid we will be able to find a solution that we both like. If we decide on having more macros, then so be it.

Can we get another opinion on this? @Bensk1 @TwoLaid @lanice @lawben @mjendruk

Reply:
You're afraid we might find consensus? :D :D

Not dead set on anything here, but the readability of tests is close to my heart. Let the council of Opossums decide.

Reply:
I don't mind having more macros, if they clearly state what they are for. I have to be honest that when reading `EXPECT_COLUMN_EXPRESSION` I have no idea what the test actually does. Maybe we could stick with a clear naming convention for such macros, i.e. `EXPECT_X_EQ` or `EXPECT_X_GT`.  

But before we start throwing around with new macros, I think it only makes sense to create them if the asserts they have are non-trivial. I see the point with `EXPECT_TABLE_EQ` and the like but a new macro just to save to one further attribute access in the test is overkill.

Reply:
When seeing an invocation "EXPECT_COLUMN_EXPRESSION" on an `Expression` and a `ColumnID`, you are NOT thinking that this tests whether the `Expression` represents the particular Column with the specified `ColumnID`? Okay... Then how would you name it so that this thought might get triggered?

Reply:
(sounds a bit more aggressive than intended, sorry :) )

Reply:
okay yea, I can't really come up with a better name wither ;) but we might still want to add an `EQ` --> `EXPECT_COLUM_EXPRESSION_EQ`. That would convey that the Expression, which we want to represent a column with, is equal to that specific column.

Reply:
"But before we start throwing around with new macros, I think it only makes sense to create them if the asserts they have are non-trivial."

I think this makes sense. If we can cover something with `EXPECT_EQ` then we should do it.

Reply:
let me just throw in that "If we can cover something with EXPECT_EQ then we should do it" doesn't apply to `EXPECT_COLUM_EXPRESSION` which actually looks like this:

``` 
if (expression->type() != ExpressionType::Column) {
    return false;
  }

  return expression->column_id() == column_id;
```

because `Expression::column_id()` intentionally Asserts that the Expression actually contains a ColumnID.

Reply:
Alright, so everybody takes issue with these new testing assertion, but so far proposed solutions leave a bit to be desired. 
- Everybody is fine with `EXPECT_TABLE_EQ`, `EXPECT_TABLE_EQ_UNORDERED` and `EXPECT_TABLE_EQ_ORDERED`, am I right?
- `EXPECT_REL_NEAR` should be a macro whose comment makes clear what its doing. I will NOT have the `std::fabs(reference_value* rel_error)` bit in my test code.
- So you take issue with `EXPECT_STORED_TABLE_NODE`, `EXPECT_COLUMN_EXPRESSION` and `EXPECT_AGGREGATE_FUNCTION_EXPRESSION`? Have you taken a look at their implementations in their respective `check_*` methods? These are not trivial one liners. Do you really want these 3-4 lines of code everywhere where one of these macros is used? Also, even if error messages leave a bit to be desired at the moment, when using these macros we can later change their implementation to generate better readable output.
- Or are you actually proposing to use `EXPECT_TRUE(check_stored_table_node(node, table_name))` instead of `EXPECT_STORED_TABLE_NODE`? See the bit about error output above.
- (I have quite a bit more of Logical Query Plan structure checking code on other branches. I very much feels using macros like these makes the tests better write- and readable.)

Reply:
bump @mrks, @TwoLaid Opinions on my comment above? At this point, just clearly tell me what you want me to do to get this PR merged ;)

Reply:
Do we want `EXPECT_TABLE_EQ`? Right now, I don't know what the default is and would have to look it up.

I still don't see why we need to add a macro for `EXPECT_COLUMN_EXPRESSION` that calls `check_column_expression` which then compares the `column_id`. How is that better than `EXPECT_EQ(projection_node->column_expressions()[1], ColumnID{4});`?

Another issue that I have with the macros is that it is unclear what they are testing. For `EXPECT_COLUMN_EXPRESSION`, it does not test the alias.

Reply:
Discussed this with @mrzzzrm in person. We have a compromise that we can both live with. Also, at least as important, it can end this discussion.

We leave the macros in, because it makes sense to not repeat the same lines over and over again. However, these macros will directly expand into a couple of ASSERT_EQ/... macros without having a function in between.